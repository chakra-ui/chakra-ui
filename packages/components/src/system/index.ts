export * from "./hooks"
export * from "./system.types"
export * from "./providers"
export * from "./system"
export * from "./forward-ref"
export * from "./use-style-config"
export * from "./factory"
export { shouldForwardProp } from "./should-forward-prop"
export { useTheme } from "./use-theme"
