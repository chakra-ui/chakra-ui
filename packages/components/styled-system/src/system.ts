import mergeWith from "lodash.mergewith"

import {
  background,
  border,
  color,
  effect,
  filter,
  flexbox,
  grid,
  interactivity,
  layout,
  list,
  others,
  position,
  ring,
  space,
  textDecoration,
  transform,
  transition,
  typography,
  scroll,
} from "./config"
import { pseudoPropNames, pseudoSelectors } from "./pseudos"

export const systemProps = mergeWith(
  {},
  background,
  border,
  color,
  flexbox,
  layout,
  filter,
  ring,
  interactivity,
  grid,
  others,
  position,
  effect,
  space,
  scroll,
  typography,
  textDecoration,
  transform,
  list,
  transition,
)

const layoutSystem = Object.assign({}, space, layout, flexbox, grid, position)
export const layoutPropNames = Object.keys(
  layoutSystem,
) as (keyof typeof layoutSystem)[]

export const propNames = [...Object.keys(systemProps), ...pseudoPropNames]
const styleProps = { ...systemProps, ...pseudoSelectors }

export const isStyleProp = (prop: string) => prop in styleProps
