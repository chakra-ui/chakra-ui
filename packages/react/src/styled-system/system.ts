import { isCssProperty } from "@pandacss/is-valid-prop"
import {
  type Dict,
  compact,
  flatten,
  isObject,
  memo,
  mergeWith,
  splitProps,
} from "../utils"
import { createBreakpoints } from "./breakpoints"
import { createConditions } from "./conditions"
import { mergeConfigs } from "./config"
import { createCssFn } from "./css"
import { createRecipeFn } from "./cva"
import { createLayers } from "./layers"
import { createNormalizeFn } from "./normalize"
import { createPreflight } from "./preflight"
import { createSerializeFn } from "./serialize"
import { createSlotRecipeFn } from "./sva"
import { createTokenDictionary } from "./token-dictionary"
import type {
  SystemConfig,
  SystemContext,
  TokenDictionary,
  TokenFn,
} from "./types"
import { createUtility } from "./utility"

export function createSystem(...configs: SystemConfig[]): SystemContext {
  const config = mergeConfigs(...configs)
  const {
    theme = {},
    utilities = {},
    globalCss = {},
    cssVarsRoot = ":where(:root, :host)",
    cssVarsPrefix = "chakra",
    preflight,
  } = config

  const layers = createLayers(config)

  const tokens = createTokenDictionary({
    breakpoints: theme.breakpoints,
    tokens: theme.tokens,
    semanticTokens: theme.semanticTokens,
    prefix: cssVarsPrefix,
  })

  const breakpoints = createBreakpoints(theme.breakpoints ?? {})

  const conditions = createConditions({
    conditions: config.conditions ?? {},
    breakpoints,
  })

  const utility = createUtility({
    config: utilities,
    tokens,
  })

  function assignComposition() {
    const { textStyles, layerStyles, animationStyles } = theme

    const compositions = compact({
      textStyle: textStyles,
      layerStyle: layerStyles,
      animationStyle: animationStyles,
    })

    for (const [key, values] of Object.entries(compositions)) {
      const flatValues = flatten(
        values ?? {},
        (v) => isObject(v) && "value" in v,
      )

      utility.register(key, {
        values: Object.keys(flatValues),
        transform(value) {
          return css(flatValues[value])
        },
      })
    }
  }

  assignComposition()
  utility.addPropertyType("animationName", Object.keys(theme.keyframes ?? {}))

  const properties = new Set(["css", ...utility.keys(), ...conditions.keys()])

  const isValidProperty = memo(
    (prop: string) => properties.has(prop) || isCssProperty(prop),
  )

  const normalizeValue = (value: any): any => {
    if (Array.isArray(value)) {
      return value.reduce((acc, current, index) => {
        const key = conditions.breakpoints[index]
        if (current != null) acc[key] = current
        return acc
      }, {})
    }
    return value
  }

  const normalizeFn = createNormalizeFn({
    utility,
    normalize: normalizeValue,
  })

  const serialize = createSerializeFn({
    conditions,
    isValidProperty,
  })

  const css = createCssFn({
    transform: utility.transform,
    conditions,
    normalize: normalizeFn,
  })

  const cva = createRecipeFn({
    css: css as any,
    conditions,
    normalize: normalizeFn,
    layers,
  })

  const sva = createSlotRecipeFn({ cva })

  function getTokenCss() {
    const result: Dict = {}

    for (const [key, values] of tokens.cssVarMap.entries()) {
      const varsObj = Object.fromEntries(values) as any
      if (Object.keys(varsObj).length === 0) continue

      if (key === "base") {
        const cssObj = css(serialize({ [cssVarsRoot]: varsObj }))
        mergeWith(result, cssObj)
      } else {
        const cssObject = css(serialize({ [key]: varsObj }))
        mergeWith(result, cssObject)
      }
    }

    return layers.wrap("tokens", result)
  }

  function getGlobalCss() {
    const keyframes = Object.fromEntries(
      Object.entries(theme.keyframes ?? {}).map(([key, value]) => [
        `@keyframes ${key}`,
        value,
      ]),
    )
    const result = Object.assign({}, keyframes, css(serialize(globalCss)))
    return layers.wrap("base", result)
  }

  function splitCssProps(props: any) {
    return splitProps(props, isValidProperty as any)
  }

  function getPreflightCss() {
    const result = createPreflight({ preflight })
    return layers.wrap("reset", result)
  }

  const tokenMap = getTokenMap(tokens)

  const tokenFn: TokenFn = (path: string, fallback?: any) => {
    return tokenMap.get(path)?.value || fallback
  }

  tokenFn.var = (path: string, fallback?: any) => {
    return tokenMap.get(path)?.variable || fallback
  }

  function getRecipe(key: string, fallback?: any) {
    return theme.recipes?.[key] ?? fallback
  }

  function getSlotRecipe(key: string, fallback?: any) {
    return theme.slotRecipes?.[key] ?? fallback
  }

  function isRecipe(key: string) {
    return Object.hasOwnProperty.call(theme.recipes ?? {}, key)
  }

  function isSlotRecipe(key: string) {
    return Object.hasOwnProperty.call(theme.slotRecipes ?? {}, key)
  }

  function hasRecipe(key: string) {
    return isRecipe(key) || isSlotRecipe(key)
  }

  return {
    $$chakra: true,
    _config: config,
    breakpoints,
    tokens,
    conditions,
    utility,
    token: tokenFn,
    properties,
    layers,
    isValidProperty,
    splitCssProps: splitCssProps as any,
    normalizeValue,
    getTokenCss,
    getGlobalCss,
    getPreflightCss,
    css: css as any,
    cva,
    sva,
    getRecipe,
    getSlotRecipe,
    hasRecipe,
    isRecipe,
    isSlotRecipe,
  }
}

function getTokenMap(tokens: TokenDictionary) {
  const map = new Map<string, { value: string; variable: string }>()

  tokens.allTokens.forEach((token) => {
    const { cssVar, virtual, conditions } = token.extensions
    const value = !!conditions || virtual ? cssVar!.ref : token.value
    map.set(token.name, { value, variable: cssVar!.ref })
  })

  return map
}

export const isValidSystem = (mod: unknown): mod is SystemContext => {
  return isObject(mod) && !!Reflect.get(mod, "$$chakra")
}
