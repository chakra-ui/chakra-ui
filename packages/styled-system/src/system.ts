import { mergeWith, objectKeys } from "@chakra-ui/utils"
import {
  background,
  border,
  color,
  effect,
  filter,
  flexbox,
  grid,
  interactivity,
  layout,
  list,
  others,
  position,
  ring,
  space,
  textDecoration,
  transform,
  transition,
  typography,
} from "./config"
import { pseudoPropNames, pseudoSelectors } from "./pseudos"

export const systemProps = mergeWith(
  {},
  background,
  border,
  color,
  flexbox,
  layout,
  filter,
  ring,
  interactivity,
  grid,
  others,
  position,
  effect,
  space,
  typography,
  textDecoration,
  transform,
  list,
  transition,
)

const layoutSystem = Object.assign({}, space, layout, flexbox, grid, position)
export const layoutPropNames = objectKeys(layoutSystem)

export const propNames = [...objectKeys(systemProps), ...pseudoPropNames]
const styleProps = { ...systemProps, ...pseudoSelectors }

export const isStyleProp = (prop: string) => prop in styleProps
