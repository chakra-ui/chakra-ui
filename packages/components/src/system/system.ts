import { mergeRefs } from "@chakra-ui/hooks/use-merge-refs"
import {
  css,
  isStylePropFn,
  StyleProps,
  SystemStyleObject,
} from "@chakra-ui/styled-system"
import {
  assignAfter,
  compact,
  interopDefault,
  runIfFn,
  splitProps,
} from "@chakra-ui/utils"
import { Dict } from "@chakra-ui/utils/types"
import createStyled, { CSSObject, FunctionInterpolation } from "@emotion/styled"
import {
  Children,
  createElement,
  ElementType,
  forwardRef,
  isValidElement,
  useMemo,
} from "react"
import { useColorMode } from "../color-mode"
import { mergeProps } from "./merge-props"
import { shouldForwardProp } from "./should-forward-prop"
import {
  AsChildProps,
  AsProps,
  ChakraComponent,
  ChakraProps,
  PropsOf,
} from "./system.types"
import { DOMElements } from "./system.utils"

const emotion_styled = interopDefault(createStyled)

type StyleResolverProps = SystemStyleObject & {
  __css?: SystemStyleObject
  sx?: SystemStyleObject
  theme: any
  css?: CSSObject
}

interface GetStyleObject {
  (options: {
    baseStyle?:
      | SystemStyleObject
      | ((props: StyleResolverProps) => SystemStyleObject)
  }): FunctionInterpolation<StyleResolverProps>
}

/**
 * Style resolver function that manages how style props are merged
 * in combination with other possible ways of defining styles.
 *
 * For example, take a component defined this way:
 * ```jsx
 * <Box fontSize="24px" sx={{ fontSize: "40px" }}></Box>
 * ```
 *
 * We want to manage the priority of the styles properly to prevent unwanted
 * behaviors. Right now, the `sx` prop has the highest priority so the resolved
 * fontSize will be `40px`
 */
export const toCSSObject: GetStyleObject =
  ({ baseStyle }) =>
  (props) => {
    const { theme, css: cssProp, __css, sx, ...restProps } = props
    const isStyleProp = isStylePropFn(theme)
    const [styleProps] = splitProps(restProps, isStyleProp)

    const finalBaseStyle = runIfFn(baseStyle, props)
    const finalStyles = assignAfter(
      {},
      __css,
      finalBaseStyle,
      compact(styleProps),
      sx,
    )
    const computedCSS = css(finalStyles)(props.theme)
    return cssProp ? [computedCSS, cssProp] : computedCSS
  }

export interface ChakraStyledOptions extends Dict {
  shouldForwardProp?(prop: string): boolean
  label?: string
  baseStyle?:
    | SystemStyleObject
    | ((props: StyleResolverProps) => SystemStyleObject)
}

export function styled<T extends ElementType, P extends object = {}>(
  component: T,
  options?: ChakraStyledOptions,
) {
  const { baseStyle, ...styledOptions } = options ?? {}

  if (!styledOptions.shouldForwardProp) {
    styledOptions.shouldForwardProp = shouldForwardProp
  }

  const styleObject = toCSSObject({ baseStyle })

  const Component = emotion_styled(
    component as React.ComponentType<any>,
    styledOptions,
  )(styleObject)

  const __el = forwardRef<any, any>(function ChakraComponent(props, ref) {
    const { asChild, children, ...restProps } = props

    const { colorMode, forced } = useColorMode()

    const dataTheme = forced ? colorMode : undefined

    if (!asChild) {
      return createElement(
        Component,
        {
          ref,
          "data-theme": dataTheme,
          ...restProps,
        },
        children,
      )
    }

    const onlyChild = Children.only(props.children)

    if (isValidElement(onlyChild)) {
      const composedProps = mergeProps(restProps, onlyChild.props ?? {})

      const composedRef = ref
        ? mergeRefs(ref, (onlyChild as any).ref)
        : (onlyChild as any).ref

      // eslint-disable-next-line react-hooks/rules-of-hooks
      const styledElement = useMemo(
        () => emotion_styled(onlyChild.type as any, styledOptions)(styleObject),
        [onlyChild.type],
      )

      return createElement(styledElement, {
        ref: composedRef,
        "data-theme": dataTheme,
        ...composedProps,
      })
    }

    return onlyChild
  })

  return __el as ChakraComponent<T, P>
}

export type HTMLChakraComponents = {
  [Tag in DOMElements]: ChakraComponent<Tag, {}>
}

export type HTMLChakraProps<T extends ElementType> = Omit<
  PropsOf<T>,
  "ref" | keyof StyleProps
> &
  ChakraProps &
  AsChildProps &
  AsProps
