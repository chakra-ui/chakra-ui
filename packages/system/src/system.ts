import {
  css,
  isStyleProp,
  StyleProps,
  SystemStyleObject,
} from "@chakra-ui/styled-system"
import { filterUndefined, objectFilter, runIfFn } from "@chakra-ui/utils"
import _styled, { CSSObject, FunctionInterpolation } from "@emotion/styled"
import { shouldForwardProp } from "./should-forward-prop"
import { As, ChakraComponent, ChakraProps, PropsOf } from "./system.types"
import { domElements, DOMElements } from "./system.utils"

type StyleResolverProps = SystemStyleObject & {
  __css?: SystemStyleObject
  sx?: SystemStyleObject
  theme: any
  css?: CSSObject
}

interface GetStyleObject {
  (options: {
    baseStyle?:
      | SystemStyleObject
      | ((props: StyleResolverProps) => SystemStyleObject)
  }): FunctionInterpolation<StyleResolverProps>
}

/**
 * Style resolver function that manages how style props are merged
 * in combination with other possible ways of defining styles.
 *
 * For example, take a component defined this way:
 * ```jsx
 * <Box fontSize="24px" sx={{ fontSize: "40px" }}></Box>
 * ```
 *
 * We want to manage the priority of the styles properly to prevent unwanted
 * behaviors. Right now, the `sx` prop has the highest priority so the resolved
 * fontSize will be `40px`
 */
export const toCSSObject: GetStyleObject = ({ baseStyle }) => (props) => {
  const { theme, css: cssProp, __css, sx, ...rest } = props
  const styleProps = objectFilter(rest, (_, prop) => isStyleProp(prop))
  const finalBaseStyle = runIfFn(baseStyle, props)
  const finalStyles = Object.assign(
    {},
    __css,
    finalBaseStyle,
    filterUndefined(styleProps),
    sx,
  )
  const computedCSS = css(finalStyles)(props.theme)
  return cssProp ? [computedCSS, cssProp] : computedCSS
}

interface StyledOptions {
  shouldForwardProp?(prop: string): boolean
  label?: string
  baseStyle?:
    | SystemStyleObject
    | ((props: StyleResolverProps) => SystemStyleObject)
}

export function styled<T extends As, P = {}>(
  component: T,
  options?: StyledOptions,
) {
  const { baseStyle, ...styledOptions } = options ?? {}

  if (!styledOptions.shouldForwardProp) {
    styledOptions.shouldForwardProp = shouldForwardProp
  }

  const styleObject = toCSSObject({ baseStyle })
  return _styled(
    component as React.ComponentType<any>,
    styledOptions,
  )(styleObject) as ChakraComponent<T, P>
}

export type HTMLChakraComponents = {
  [Tag in DOMElements]: ChakraComponent<Tag, {}>
}

export type HTMLChakraProps<T extends As> = Omit<
  PropsOf<T>,
  T extends "svg"
    ? "ref" | "children" | keyof StyleProps
    : "ref" | keyof StyleProps
> &
  ChakraProps & { as?: As }

type ChakraFactory = {
  <T extends As, P = {}>(
    component: T,
    options?: StyledOptions,
  ): ChakraComponent<T, P>
}

export const chakra = (styled as unknown) as ChakraFactory &
  HTMLChakraComponents

domElements.forEach((tag) => {
  chakra[tag] = chakra(tag)
})
