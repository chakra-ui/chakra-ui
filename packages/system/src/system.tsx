import { parser as systemProps } from "@chakra-ui/parser"
import createStyled from "./create-styled"
import { As, ChakraComponent, Options } from "./system.types"
import {
  applyProp,
  componentProps,
  domElements,
  DOMElements,
  pseudoProps,
  truncateProp,
  layerStyleProp,
} from "./system.utils"

function styled<T extends As, P = {}>(component: T, options?: Options<T, P>) {
  return createStyled<T, P>(component, options)(
    componentProps(options),
    systemProps,
    pseudoProps,
    applyProp(component),
    truncateProp,
    layerStyleProp,
  )
}

type ChakraJSXElements = {
  [Tag in DOMElements]: ChakraComponent<Tag, {}>
}

type CreateChakraComponent = {
  /**
   * The styled function and object that allows to create a
   * functional component that can receive chakra's style props.
   *
   * @param component - The base component or tag to render
   * @param baseStyle - The initial styles to apply (accepts chakra styles)
   * @param attrs - The initial props to the component (valid html or component props)
   */
  <T extends As, P = {}>(
    component: T,
    options?: Options<T, P>,
  ): ChakraComponent<T, P>
}

export const chakra = (styled as unknown) as CreateChakraComponent &
  ChakraJSXElements

domElements.forEach((tag) => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  chakra[tag] = chakra(tag)
})
