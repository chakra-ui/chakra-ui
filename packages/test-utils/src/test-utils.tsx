import theme from "@chakra-ui/preset-base"
import { GlobalStyle, ThemeProvider } from "@chakra-ui/system"
import CSSReset from "@chakra-ui/css-reset"
import "@testing-library/jest-dom/extend-expect"
import { render, RenderOptions } from "@testing-library/react"
import * as React from "react"
import { toHaveNoViolations } from "jest-axe"

expect.extend(toHaveNoViolations)

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <ThemeProvider theme={theme}>
    <CSSReset />
    <GlobalStyle />
    {children}
  </ThemeProvider>
)

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options })

export * from "@testing-library/react"

export {act as invoke, renderHook} from "@testing-library/react-hooks"
export type  { RenderHookOptions, RenderHookResult} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export { customRender as render }

export * from "jest-axe"
