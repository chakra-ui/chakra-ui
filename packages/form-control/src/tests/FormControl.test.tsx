import * as React from "react"
import { chakra, PropsOf } from "@chakra-ui/system"
import { render, fireEvent } from "@chakra-ui/test-utils"
import {
  FormControl,
  FormErrorIcon,
  FormErrorMessage,
  FormHelperText,
  FormLabel,
  RequiredIndicator,
  useField,
  ControlProps,
} from "../FormControl"

type OmittedTypes = "disabled" | "required" | "readOnly"
type InputProps = Omit<PropsOf<typeof StyledInput>, OmittedTypes> & ControlProps

// Create an input that consumes useField
const StyledInput = chakra<
  "input",
  { focusBorderColor?: string; errorBorderColor?: string }
>("input", { themeKey: "Input" })

const Input = React.forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const inputProps = useField<HTMLInputElement>(props)
  return <StyledInput ref={ref} {...inputProps} />
})

test("FormControl renders correctly in default state", () => {
  const { asFragment } = render(
    <FormControl id="name">
      <FormLabel>Name</FormLabel>
      <Input placeholder="Name" />
      <FormHelperText>Enter your name please!</FormHelperText>
      <FormErrorMessage>Your name is invalid</FormErrorMessage>
    </FormControl>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("FormControl renders correctly when required", () => {
  const { asFragment } = render(
    <FormControl id="name" isRequired>
      <FormLabel>Name</FormLabel>
      <RequiredIndicator />
      <Input placeholder="Name" />
      <FormHelperText>Enter your name please!</FormHelperText>
      <FormErrorMessage>Your name is invalid</FormErrorMessage>
    </FormControl>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("FormControl renders correctly when invalid", () => {
  const { asFragment } = render(
    <FormControl id="name" isInvalid>
      <FormLabel>Name</FormLabel>
      <RequiredIndicator />
      <Input placeholder="Name" />
      <FormHelperText>Enter your name please!</FormHelperText>
      <FormErrorIcon />
      <FormErrorMessage>Your name is invalid</FormErrorMessage>
    </FormControl>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("useField calls provided input callbacks", () => {
  const onFocus = jest.fn()
  const onBlur = jest.fn()

  const utils = render(
    <FormControl id="name">
      <FormLabel>Name</FormLabel>
      <Input
        data-testid="input"
        placeholder="Name"
        onFocus={onFocus}
        onBlur={onBlur}
      />
    </FormControl>,
  )
  const input = utils.getByTestId("input")

  fireEvent.focus(input)
  fireEvent.blur(input)
  expect(onFocus).toHaveBeenCalled()
  expect(onBlur).toHaveBeenCalled()
})

test("has the proper aria attributes", async () => {
  const utils = render(
    <FormControl id="name">
      <FormLabel>Name</FormLabel>
      <Input placeholder="Name" />
      <FormHelperText>Enter your name please!</FormHelperText>
    </FormControl>,
  )
  let input = utils.getByLabelText("Name")

  expect(input).toHaveAttribute("aria-describedby", "name-helptext")
  expect(input).not.toHaveAttribute("aria-invalid")
  expect(input).not.toHaveAttribute("aria-required")
  expect(input).not.toHaveAttribute("aria-readonly")

  utils.rerender(
    <FormControl id="name" isRequired isInvalid isReadOnly>
      <FormLabel>Name</FormLabel>
      <RequiredIndicator data-testid="indicator" />
      <Input placeholder="Name" />
      <FormHelperText>Enter your name please!</FormHelperText>
      <FormErrorMessage data-testid="error">
        Your name is invalid
      </FormErrorMessage>
    </FormControl>,
  )
  input = utils.getByLabelText("Name")
  const indicator = utils.getByTestId("indicator")
  const errorMessage = utils.getByTestId("error")

  expect(input).toHaveAttribute("aria-invalid", "true")
  expect(input).toHaveAttribute("aria-required", "true")
  expect(input).toHaveAttribute("aria-readonly", "true")
  expect(input).toHaveAttribute(
    "aria-describedby",
    "name-feedback name-helptext",
  )
  expect(indicator).toHaveAttribute("aria-hidden")
  expect(errorMessage).toHaveAttribute("aria-live", "polite")
})

test("has the correct role attributes", () => {
  const utils = render(
    <FormControl data-testid="control" id="name" isRequired>
      <FormLabel>Name</FormLabel>
      <RequiredIndicator data-testid="indicator" />
      <Input placeholder="Name" />
    </FormControl>,
  )
  const control = utils.getByTestId("control")
  const indicator = utils.getByTestId("indicator")

  expect(utils.getByRole("presentation", { hidden: true })).toEqual(indicator)
  expect(utils.getByRole("group")).toEqual(control)
})

test("has the correct data attributes", () => {
  const utils = render(
    <FormControl
      data-testid="control"
      id="name"
      isRequired
      isInvalid
      isDisabled
      isLoading
      isReadOnly
    >
      <FormLabel data-testid="label">Name</FormLabel>
      <RequiredIndicator data-testid="indicator" />
      <Input placeholder="Name" />
      <FormHelperText data-testid="helper-text">
        Please enter your name!
      </FormHelperText>
      <FormErrorMessage data-testid="error-message">
        Your name is invalid.
      </FormErrorMessage>
    </FormControl>,
  )
  const control = utils.getByTestId("control")
  const label = utils.getByTestId("label")
  const indicator = utils.getByTestId("indicator")
  const helperText = utils.getByTestId("helper-text")
  const errorMessage = utils.getByTestId("error-message")

  fireEvent.focus(utils.getByLabelText("Name"))

  expect(control).toHaveAttribute("data-chakra-form-control")
  expect(label).toHaveAttribute("data-chakra-form-label")
  expect(label).toHaveAttribute("data-focus")
  expect(label).toHaveAttribute("data-invalid")
  expect(label).toHaveAttribute("data-loading")
  expect(label).toHaveAttribute("data-readonly")
  expect(indicator).toHaveAttribute("data-chakra-required-indicator")
  expect(helperText).toHaveAttribute("data-chakra-form-helper-text")
  expect(errorMessage).toHaveAttribute("data-chakra-form-error-message")
})
