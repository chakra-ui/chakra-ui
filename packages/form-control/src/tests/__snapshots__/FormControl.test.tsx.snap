// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControl renders correctly in default state 1`] = `
<DocumentFragment>
  <div
    class="css-79elbk"
    data-chakra-form-control=""
    role="group"
  >
    <label
      class="css-1q2bw9z"
      data-chakra-form-label=""
      for="name"
      id="name-label"
    >
      Name
    </label>
    <input
      aria-describedby="name-helptext"
      class="css-8b1ui0"
      id="name"
      placeholder="Name"
    />
    <div
      class="css-1qoocny"
      data-chakra-form-helper-text=""
      id="name-helptext"
    >
      Enter your name please!
    </div>
  </div>
</DocumentFragment>
`;

exports[`FormControl renders correctly when invalid 1`] = `
<DocumentFragment>
  <div
    class="css-79elbk"
    data-chakra-form-control=""
    role="group"
  >
    <label
      class="css-1q2bw9z"
      data-chakra-form-label=""
      data-invalid=""
      for="name"
      id="name-label"
    >
      Name
    </label>
    <input
      aria-describedby="name-feedback name-helptext"
      aria-invalid="true"
      class="css-8b1ui0"
      id="name"
      placeholder="Name"
    />
    <div
      class="css-1qoocny"
      data-chakra-form-helper-text=""
      id="name-helptext"
    >
      Enter your name please!
    </div>
    <svg
      aria-hidden="true"
      class="css-1gn1kov"
      data-form-error-icon=""
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
        fill="currentColor"
      />
    </svg>
    <div
      aria-live="polite"
      class="css-v1r939"
      data-chakra-form-error-message=""
      id="name-feedback"
    >
      Your name is invalid
    </div>
  </div>
</DocumentFragment>
`;

exports[`FormControl renders correctly when required 1`] = `
<DocumentFragment>
  <div
    class="css-79elbk"
    data-chakra-form-control=""
    role="group"
  >
    <label
      class="css-1q2bw9z"
      data-chakra-form-label=""
      for="name"
      id="name-label"
    >
      Name
    </label>
    <span
      aria-hidden="true"
      class="css-i02c70"
      data-chakra-required-indicator=""
      role="presentation"
    />
    <input
      aria-describedby="name-helptext"
      aria-required="true"
      class="css-8b1ui0"
      id="name"
      placeholder="Name"
    />
    <div
      class="css-1qoocny"
      data-chakra-form-helper-text=""
      id="name-helptext"
    >
      Enter your name please!
    </div>
  </div>
</DocumentFragment>
`;
